### DISCUSSION TITLE
RFC: Make rule classes internal

### DISCUSSION THREAD

**Discussion by jozsefdamokos** [shopware/platform#5785] [RFC / Proposal]
## Summary
As a first step of simplifying our rule system and allowing us to easier improve it in the future this RFC proposes making rule classes internal and encouraging extension only through new rule classes. These changes aim to streamline maintenance, improve extensibility, and accelerate future evolution without altering current behaviour.

---

## Motivation
The current rule system is flexible but complex, making it difficult to evolve and maintain. Restricting access to internal rule classes and providing a clearer path for extensions will reduce misuse, improve maintainability, and enable faster improvements.

---

### Changes
1. **Internalize Rule Classes**  
   - Mark rule classes as internal to limit direct usage by third parties.  
2. **Extend via New Rule Classes**  
   - Developers should create new rule classes for custom logic rather than modifying existing ones.  
3. **Analyze Rule Class Extensions**  
   - Assess how often third parties currently extend rule classes to inform and communicate the transition. If you already extend any of the current rule classes please comment your scenario and assess whether the same result can be achieved in alternative ways.

### Benefits
- Faster evolution of the rule system.
- Clearer, more robust extension mechanisms for developers.
- Easier maintenance and reduced coupling.

---

## Compatibility and Migration
- **Database Behavior**: No changes; the existing rules and the type remains stored in the database as-is. Because this is the case their behaviour cannot and should not change. Only their internals can change. For example a rule that had a string constraint could now have a list of strings as constraint. However, in order to support the old behaviour, a database migration is needed to translate the previous configuration to the new one.

**Comment by schneider-felix** [comment]
I have never extended the rule classes in any project. The way for me has always been to add additional rules. I think extending the rules is a really uncommon usecase. I think this change would be fine.

**Comment by aragon999** [comment]
I had two cases (I can now remember top of my head) where I extended rules:
1. Custom [`LineItemOfTypeRule`](https://github.com/shopware/shopware/blob/619f62c6b2f2590aef73479ed2c9a68b7eb29945/src/Core/Checkout/Cart/Rule/LineItemOfTypeRule.php#L14)-types (in that case three additional types have been added)
2. Additional matching of ZIP codes

I think in particular in the first case, when not be able to add this to the existing rule, it might be more confusing for the Shop administrator, which rule is needed. Either we could add three different rules for each type, or add a rule which is doing the exactly the same as the current and the Shop administrator needs to know which rule he should use. 

The same is true for the ZIP code rules.

**Reply by jozsefdamokos** [reply]
Indeed, that could be confusing, having the original rule and the new rule which is basically an extension of it. What about limiting the possibilities of the admin user, like hiding the original rule from the selection. Would that be a possibility?

**Reply by aragon999** [reply]
I mean that probably is also not ideal, as only one plugin can add types then. 

Also this approach may be more error prone, as the user comes to the issue tracker, reports a problem, as the rule is named the same, it wont work, but the error is actually in the implementation of the copied rule. 

Also existing rules have problems in that case.

**Reply by jozsefdamokos** [reply]
Yes, you're right. That is not really an option then

**Comment by riconeitzel** [comment]
Will the new system be open for decorations instead?

**Reply by jozsefdamokos** [reply]
The rule classes themselves not in this case. Decoration would require them to remain public.

**Comment by OliverSkroblin** [comment]
it would also be awesome to remove the different rule scopes and unify them. 

![image](https://github.com/user-attachments/assets/cc66da60-65df-40d7-a455-68b29de252a3)

I guess Cart and Checkout can be merged or not?

**Reply by jozsefdamokos** [reply]
Yes, that is a good idea for next step

**Comment by jozsefdamokos** [comment]
See the related PR together with and ADR here: https://github.com/shopware/shopware/pull/6448

**Comment by tinect** [comment]
How would I use existing Rules for PriceDefinitions filter? While this is now yelling:
![image](https://github.com/user-attachments/assets/bfefda24-d20f-4600-8f2f-434478c2b4e1)

**Reply by jozsefdamokos** [reply]
Thanks for raising this. We will make it final instead of internal

**Comment by OliverSkroblin** [comment]
Wouldn't be the correct approach to mark them as final instead of internal?

When i understand the ADR correctly, you want to prevent that other developers extends your rules - legit. 

But when you mark this as @internal - i guess you running into BC problems. Because the properties of the __construct are kind of public API because the data for it is in the database AND, you need to provide this via API calls to create this rules. Changing such __construct parameters, would lead to a BC or not?

**Reply by jozsefdamokos** [reply]
Good point, making them final instead of internal is the only way

**Comment by augsteyer** [comment]
What happens if I want to re-use all cart item rules with a slightly different matching mechanism?

For example, I want an event trigger when an item gets removed from cart. I want to use the cart item rules, but with slight modification that I am not processing all/any of the items in the cart, but only the one being removed. So I would like to extend every cart item rule, and modify the matcher function to work only on the "removed" product instead of "all" or "any" of them. The finality of the class will force me to copy all of the 33 Cart item rule classes & maintain them with any change to those classes.

There is another approach to it, and it's rewriting the container when selecting the "all" or "any" drop box. Just add another option, like "On Remove". However, there will still be a problem when I will be trying to exclude these from being used in the promotions via `ruleConditionService.addAwarenessConfiguration`. Since I am not extending/creating the cart item rules, but using default ones with a different matcher container (onRemove), I cannot add an awareness config to restrict it (otherwise, I will be restricting the all/any matcher rules too).

So we are back to copy/pasting all the cart item rules, right? 

P.S. hope that makes any sense.
